README.md
# pygltflib

This is a library for reading, writing and handling GLTF files. It works for Python3.6 and above.

It supports the entire specification, including materials and animations. Main features are:
* GLB and GLTF support
* Buffer data conversion
* Extensions
* All attributes are type-hinted

# Table of Contents

* [Quickstart](#quickstart)
  * [Install](#install)
  * [How do I...](#how-do-i)
    * [Create an empty GLTF2 object?](#create-an-empty-gltf2-object)
    * [Add a scene?](#add-a-scene)
    * [Load a file?](#load-a-file)
    * [Load a binary GLB file?](#load-a-binary-glb-file)
    * [Load a binary file with an unusual extension?](#load-a-binary-file-with-an-unusual-extension)
    * [Access the first node (the objects comprising the scene) of a scene?](#access-the-first-node-the-objects-comprising-the-scene-of-a-scene)
    * [Create a mesh?](#create-a-mesh)
    * [Convert buffers to GLB binary buffers?](#convert-buffers-to-glb-binary-buffers)
    * [Convert buffer to data uri (embedded) buffer?](#convert-buffer-to-data-uri-embedded-buffer)
    * [Convert buffers to binary file (external) buffers?](#convert-buffers-to-binary-file-external-buffers)
    * [Convert a glb to a gltf file?](#convert-a-glb-to-a-gltf-file)
    * [Access an extension?](#access-an-extension)
    * [Add a custom attribute to Attributes?](#add-a-custom-attribute-to-attributes)
    * [Remove a bufferView?](#remove-a-bufferview)
    * [Validate a gltf object?](#validate-a-gltf-object)
    * [Convert texture images inside a GLTF file to their own PNG files?](#convert-texture-images-inside-a-gltf-file-to-their-own-png-files)
    * [Convert texture images from a GLTF file to their own PNG files using custom file names?](#convert-texture-images-from-a-gltf-file-to-their-own-png-files-using-custom-file-names)
    * [Specify a path to my images when converting to files?](#specify-a-path-to-my-images-when-converting-to-files)
    * [Export images from the GLTF file to any location (ie outside the GLTF file)?](#export-images-from-the-GLTF-file-to-any-location-ie-outside-the-GLTF-file)
    * [Import PNG files as textures into a GLTF?](#import-png-files-as-textures-into-a-gltf)
* [About](#about)
  * [Roadmap](#roadmap)
  * [Contributors](#contributors)
  * [Thanks](#thanks)
  * [Changelog](#changelog)
  * [Installing](#installing)
  * [Source](#source)
* [More Detailed Usage](#more-detailed-usage)
  * [A simple mesh](#a-simple-mesh)
  * [Reading vertex data from a primitive and/or getting bounding sphere](#reading-vertex-data-from-a-primitive-andor-getting-bounding-sphere)
  * [Create a mesh, convert to bytes, convert back to mesh](#create-a-mesh-convert-to-bytes-convert-back-to-mesh)
  * [Loading and saving](#loading-and-saving)
  * [Converting files](#converting-files)
  * [Converting buffers](#converting-buffers)
  * [Converting texture images](#converting-texture-images)
  * [Extensions](#extensions)
* [Running the tests](#running-the-tests)

## Quickstart

### Install

```
pip install pygltflib
```


### How do I...


#### Create an empty GLTF2 object?

```python
from pygltflib import GLTF2

gltf = GLTF2()
```

#### Add a scene?

```python
from pygltflib import GLTF2, Scene

gltf = GLTF2()
scene = Scene()
gltf.scenes.append(scene)  # scene available at gltf.scenes[0]
```

#### Load a file?

```python
filename = "glTF-Sample-Models/2.0/AnimatedCube/glTF/AnimatedCube.gltf"
gltf = GLTF2().load(filename)
```

#### Load a binary GLB file?

```python
glb_filename = "glTF-Sample-Models/2.0/Box/glTF-Binary/Box.glb"
glb = GLTF2().load(glb_filename)  # load method auto detects based on extension
```

#### Load a binary file with an unusual extension?

```python
glb = GLTF2().load_binary("BinaryGLTF.glk")   # load_json and load_binary helper methods
```


#### Access the first node (the objects comprising the scene) of a scene?

```python
gltf = GLTF2().load("glTF-Sample-Models/2.0/Box/glTF/Box.gltf")
current_scene = gltf.scenes[gltf.scene]
node_index = current_scene.nodes[0]  # scene.nodes is the indices, not the objects 
box = gltf.nodes[node_index]
box.matrix  # will output vertices for the box object
```


#### Create a mesh?
Consult the longer examples in the second half of this document
  * [A simple mesh](#a-simple-mesh)
  * [Reading vertex data from a primitive and/or getting bounding sphere](#reading-vertex-data-from-a-primitive-andor-getting-bounding-sphere)
  * [Create a mesh, convert to bytes, convert back to mesh](#create-a-mesh-convert-to-bytes-convert-back-to-mesh)


#### Convert buffers to glb binary buffers?

```python
from pygltflib import GLTF2, BufferFormat

gltf = GLTF2().load("glTF-Sample-Models/2.0/Box/glTF/Box.gltf")
gltf.convert_buffers(BufferFormat.BINARYBLOB)   # convert buffers to GLB blob
```

#### Convert buffer to data uri (embedded) buffer?
```python
gltf.convert_buffers(BufferFormat.DATAURI)  # convert buffer URIs to data.
```

#### Convert buffers to binary file (external) buffers?
```python
gltf.convert_buffers(BufferFormat.BINFILE)   # convert buffers to files
gltf.save("test.gltf")  # all the buffers are saved in 0.bin, 1.bin, 2.bin.
```


#### Convert a glb to a gltf file?
```python
from pygltflib.utils import glb2gltf, gltf2glb

# convert glb to gltf
glb2gltf("glTF-Sample-Models/2.0/Box/glTF-Binary/Box.glb")
```

#### Access an extension?
```python
# on a primitve
gltf.meshes[0].primitives[0].extensions['KHR_draco_mesh_compression']

# on a material
gltf.materials[0].extensions['ADOBE_materials_thin_transparency']

```

#### Add a custom attribute to Attributes?
```python
# Application-specific semantics must start with an underscore, e.g., _TEMPERATURE.
a = Attributes()
a._MYCUSTOMATTRIBUTE = 123

gltf.meshes[0].primitives[0].attributes._MYOTHERATTRIBUTE = 456
```

#### Remove a bufferView?
```python
gltf.remove_bufferView(0)  # this will update all accessors, images and sparse accessors to remove the first bufferView
```

#### Validate a gltf object?
```python
from pygltflib import GLTF2
from pygltflib.validator import validate, summary
filename = "glTF-Sample-Models/2.0/AnimatedCube/glTF/AnimatedCube.gltf"
gltf = GLTF2().load(filename)
validate(gltf)  # will throw an error depending on the problem
summary(gltf)  # will pretty print human readable summary of errors
# NOTE: Currently this experimental validator only validates a few rules about GLTF2 objects
```

#### Convert texture images inside a GLTF file to their own PNG files?
```python
from pygltflib import GLTF2
from pygltflib.utils import ImageFormat
filename = "glTF-Sample-Models/2.0/AnimatedCube/glTF/AnimatedCube.gltf"
gltf = GLTF2().load(filename)
gltf.convert_images(ImageFormat.FILE)
gltf.images[0].uri  # will now be 0.png and the texture image will be saved in 0.png
```

#### Convert texture images from a GLTF file to their own PNG files using custom file names?
```python
from pygltflib import GLTF2
from pygltflib.utils import ImageFormat
filename = "glTF-Sample-Models/2.0/AnimatedCube/glTF/AnimatedCube.gltf"
gltf = GLTF2().load(filename)
gltf.images[0].name = "cube.png"  # will save the data uri to this file (regardless of data format)
gltf.convert_images(ImageFormat.FILE)
gltf.images[0].uri  # will now be cube.png and the texture image will be saved in cube.png
```

#### Specify a path to my images when converting to files?
By default pygltflib will load images from the same location as the GLTF file.

It will also try and save image files to the that location when converting image buffers or data uris.

You can override the load/save location using the 'path' argument to `convert_images`
```python
from pygltflib import GLTF2
from pygltflib.utils import ImageFormat
filename = "glTF-Sample-Models/2.0/AnimatedCube/glTF/AnimatedCube.gltf"
gltf = GLTF2().load(filename)
gltf.images[0].name = "cube.png"  # will save the data uri to this file (regardless of data format)
gltf.convert_images(ImageFormat.FILE, path='/destination/') 
gltf.images[0].uri  # will now be cube.png and the texture image will be saved in /destination/cube.png
```


#### Export images from the GLTF file to any location (ie outside the GLTF file)?
```python
from pygltflib import GLTF2
from pygltflib.utils import ImageFormat
filename = "glTF-Sample-Models/2.0/AnimatedCube/glTF/AnimatedCube.gltf"
gltf = GLTF2().load(filename)
gltf.export_image(0, "output/cube.png", override=True)  # There is now an image file at output/cube.png
```


#### Import PNG files as textures into a GLTF?
```python
from pygltflib import GLTF2
from pygltflib.utils import ImageFormat, Image
gltf = GLTF2()
image = Image()
image.uri = "myfile.png"
gltf.images.append(image)
gltf.convert_images(ImageFormat.DATAURI)
gltf.images[0].uri  # will now be something like "data:image/png;base64,iVBORw0KGg..."
gltf.images[0].name  # will be myfile.png
```


### More Detailed Usage Below

## About
This is an unofficial library that tracks the [official file format](https://github.com/KhronosGroup/glTF/blob/master/specification/2.0/README.md) for GLTF2. 

The library was initially built to load and save simple meshes but support for the entire spec, including materials 
and animations is pretty good. Supports both json (.gltf) and binary (.glb) file formats, although .glb support 
is missing some features at the moment. 

It requires python 3.6 and above because it uses dataclasses and all attributes are type hinted. And f-strings, plenty of f-strings.

Check the table below for an idea of which sample models validate.

Questions? Contributions? Bug reports? Open an issue on the [gitlab page for the project](https://gitlab.com/dodgyville/pygltflib).
We are very interested in hearing your use cases for `pygltflib` to help drive the roadmap.

### Roadmap
* Add helper functions for creating meshes
* Test coverage
* Enforce single underscore on custom Attribute attributes 
* Investigate creating classes from extensions
* Automated validation and visual inspection

### Contributors
* Luke Miller
* Sebastian Höffner
* Arthur van Hoff
* Arifullah Jan
* Daniel Haehn
* Jon Time
* Laurie O
* Peter Suter
* Frédéric Devernay
* Julian Stirling
* Johannes Deml
* Margarida Silva 
* Patiphan Wongklaew
* Alexander Druz
* Adriano Martins
* Dzmitry Stabrouski
* irtimir
* Florian Bruggisser

#### Thanks
`pyltflib` made for 'The Beat: A Glam Noir Game' supported by Film Victoria. 

### Changelog
* 1.15.1:
  * Dataclasses install only required on python 3.6.x (cherry-pick from Saeid Akbari branch)
  * Removed deprecated `AlphaMode` after two years (use the `pygltflib.BLEND`, `pygltflib.MASK`, `pygltflib.OPAQUE` constants directly)
  * Removed deprecated `SparseAccessor` after two years (use `AccessorSparseIndices` and `AccessorSparseValues` instead)
  * Removed deprecated `MaterialTexture` after two years (use `TextureInfo` instead)
  * removed `deprecated` requirement from project

* 1.15.0: 
  * Significantly improved `save_to_bytes` performance (20x faster) (Florian Bruggisser)
    * NOTE: Underlying binary blob is now mutable instead of immutable. 

* 1.14.7
  * add `GLTF.get_data_from_buffer_uri` helper method to simplify access to buffer data (see bounding box example in README.md) (el_flamenco)

* 1.14.6
  * use compact json when saving binary glb files (Laubeee)

* 1.14.5
  * unquote filepath in compliance with standard (irtimir)

* 1.14.4
  * Add `GLTF.export_image` method to export images from an GLTF2 file to any location (Khac Hoa Le)
  * remove extraneous print message when loading extensions (Michael Daw)

See [CHANGELOG.md] (https://gitlab.com/dodgyville/pygltflib/-/blob/master/CHANGELOG.md) for older versions

## Installing
```
pip install pygltflib
```
or
```
py -m pip install pygltflib
```


## Source

```
git clone https://gitlab.com/dodgyville/pygltflib
```


## More Detailed Usage
Note: These examples use the official [sample models](https://github.com/KhronosGroup/glTF-Sample-Models) provided by Khronos at:

https://github.com/KhronosGroup/glTF-Sample-Models

### A simple mesh
```python
from pygltflib import *

# create gltf objects for a scene with a primitive triangle with indexed geometry
gltf = GLTF2()
scene = Scene()
mesh = Mesh()
primitive = Primitive()
node = Node()
buffer = Buffer()
bufferView1 = BufferView()
bufferView2 = BufferView()
accessor1 = Accessor()
accessor2 = Accessor()

# add data
buffer.uri = "data:application/octet-stream;base64,AAABAAIAAAAAAAAAAAAAAAAAAAAAAIA/AAAAAAAAAAAAAAAAAACAPwAAAAA="
buffer.byteLength = 44

bufferView1.buffer = 0
bufferView1.byteOffset = 0
bufferView1.byteLength = 6
bufferView1.target = ELEMENT_ARRAY_BUFFER

bufferView2.buffer = 0
bufferView2.byteOffset = 8
bufferView2.byteLength = 36
bufferView2.target = ARRAY_BUFFER

accessor1.bufferView = 0
accessor1.byteOffset = 0
accessor1.componentType = UNSIGNED_SHORT
accessor1.count = 3
accessor1.type = SCALAR
accessor1.max = [2]
accessor1.min = [0]

accessor2.bufferView = 1
accessor2.byteOffset = 0
accessor2.componentType = FLOAT
accessor2.count = 3
accessor2.type = VEC3
accessor2.max = [1.0, 1.0, 0.0]
accessor2.min = [0.0, 0.0, 0.0]

primitive.attributes.POSITION = 1
node.mesh = 0
scene.nodes = [0]

# assemble into a gltf structure
gltf.scenes.append(scene)
gltf.meshes.append(mesh)
gltf.meshes[0].primitives.append(primitive)
gltf.nodes.append(node)
gltf.buffers.append(buffer)
gltf.bufferViews.append(bufferView1)
gltf.bufferViews.append(bufferView2)
gltf.accessors.append(accessor1)
gltf.accessors.append(accessor2)

# save to file
gltf.save("triangle.gltf")
```


### Reading vertex data from a primitive and/or getting bounding sphere
```python
import pathlib
import struct

import miniball
import numpy
from pygltflib import GLTF2

# load an example gltf file from the khronos collection
fname = pathlib.Path("glTF-Sample-Models/2.0/Box/glTF-Embedded/Box.gltf")
gltf = GLTF2().load(fname)

# get the first mesh in the current scene (in this example there is only one scene and one mesh)
mesh = gltf.meshes[gltf.scenes[gltf.scene].nodes[0]]

# get the vertices for each primitive in the mesh (in this example there is only one)
for primitive in mesh.primitives:

    # get the binary data for this mesh primitive from the buffer
    accessor = gltf.accessors[primitive.attributes.POSITION]
    bufferView = gltf.bufferViews[accessor.bufferView]
    buffer = gltf.buffers[bufferView.buffer]
    data = gltf.get_data_from_buffer_uri(buffer.uri)

    # pull each vertex from the binary buffer and convert it into a tuple of python floats
    vertices = []
    for i in range(accessor.count):
        index = bufferView.byteOffset + accessor.byteOffset + i*12  # the location in the buffer of this vertex
        d = data[index:index+12]  # the vertex data
        v = struct.unpack("<fff", d)   # convert from base64 to three floats
        vertices.append(v)
        print(i, v)

# convert a numpy array for some manipulation
S = numpy.array(vertices)

# use a third party library to perform Ritter's algorithm for finding smallest bounding sphere
C, radius_squared = miniball.get_bounding_ball(S)

# output the results
print(f"center of bounding sphere: {C}\nradius squared of bounding sphere: {radius_squared}")
```


### Create a mesh, convert to bytes, convert back to mesh
The geometry is derived from [glTF 2.0 Box Sample](https://github.com/KhronosGroup/glTF-Sample-Models/tree/master/2.0/Box), but point normals were removed and points were reused where it was possible in order to reduce the size of the example. Be aware that some parts are hard-coded (types and shapes for en- and decoding of arrays, no bytes padding).
```python
import numpy as np
import pygltflib
```
Define mesh using `numpy`:
```python
points = np.array(
    [
        [-0.5, -0.5, 0.5],
        [0.5, -0.5, 0.5],
        [-0.5, 0.5, 0.5],
        [0.5, 0.5, 0.5],
        [0.5, -0.5, -0.5],
        [-0.5, -0.5, -0.5],
        [0.5, 0.5, -0.5],
        [-0.5, 0.5, -0.5],
    ],

CHANGELOG.md
* 1.15.1:
  * Dataclasses install only required on python 3.6.x (cherry-pick from Saeid Akbari branch)
  * Removed deprecated `AlphaMode` after two years (use the `pygltflib.BLEND`, `pygltflib.MASK`, `pygltflib.OPAQUE` constants directly)
  * Removed deprecated `SparseAccessor` after two years (use `AccessorSparseIndices` and `AccessorSparseValues` instead)
  * Removed deprecated `MaterialTexture` after two years (use `TextureInfo` instead)
  * removed `deprecated` requirement from project

* 1.15.0: 
  * Significantly improved `save_to_bytes` performance (20x faster) (Florian Bruggisser)
    * NOTE: Underlying binary blob is now mutable instead of immutable. 

* 1.14.7
  * add `GLTF.get_data_from_buffer_uri` helper method to simplify access to buffer data (see bounding box example in README.md) (el_flamenco)

* 1.14.6
  * use compact json when save binary glb files (Laubeee)

* 1.14.5
  * unquote filepath in compliance with standard (irtimir)

* 1.14.4
  * Add `GLTF.export_image` method to export images from an GLTF2 file to any location (Khac Hoa Le)
  * remove extraneous print message when loading extensions (Michael Daw)

* 1.14.3
  * add ability to save data directly in the uri field to `save_to_bytes` (Israel)
  * fix issue where attributes field is shared between two instances of Primitive (Konstantin Sinitsyn)

* 1.14.2
  * improve mimetype handling (Gabriel Unmüßig)

* 1.14.1
  * fix unicode error in setup.py (Andrew Stockton)

* 1.14.0
   * NOTE: Converting image.bufferView to image file now obeys "override" flag and also uses object path
   * fix issue where image.bufferView with value '0' is interpreted as false (Dzmitry Stabrouski)
   * fix issue where override flag ignored when converting image.bufferView to image file
   * change image.bufferView export to image path handling to be same as image.uri export
   * add longer example of mesh->bytes->mesh (Alexander Druz)
   * add GLTF.convert_images takes a 'path' argument that overrides the origin or destination
     path when convert from or to files.

* 1.13.10
   * NOTE: `GLTF2.load` now throws `FileNotFoundError` instead of failing silently on missing file.
   * fix issue where extensions with empty but valid dicts were not saving 
   * add `GLTF2.set_binary_blob` for improved access of binary data
   * split `GLTF2.save_binary` into two methods
     * GLTF2.save_binary - functions the same way as the original method
     * GLTF2.save_to_bytes - returns an array containing a binary GLTF file in an array structure 

* 1.13.9
   * split `GLTF2.load_binary` into more useful class methods (Patiphan Wongklaew)
      * GLTF2.load_binary - functions the same way as the original method
      * GLTF2.load_from_bytes - takes raw bytes directly
      * GLTF2.load_binary_from_file_object - loads from a file-like object
    * add missing test image

* 1.13.8
   * removed deprecated encoding argument from json parsers (deprecated since python 3.1)
   * add support for python 3.9 (Margarida Silva)

* 1.13.7
    * change docs and code to be more readable
* 1.13.6
    * add 4th array element to `baseColorFactor` to match specification
* 1.13.5
    * fix Matrix, Translation, Scale, and Rotation to default to None
    * change `utils.validator` by moving to `validator.validate`
    * add `validator.summary` to provide human readable output of validator
    * add some more unit tests

* 1.13.4
    * add warning to `remove_bufferView` if leaving dangling references to removed bufferView
    * add tests for `remove_bufferView`
    * add node access and remove_bufferView examples to README
    * add validation for animation channels
    * change `metallicRoughnessTexture` to use `TextureInfo`
    * change requirements to not install dataclasses in python 3.7 and above
    * fix `remove_bufferView` when sparse accessor is None
    * fix spelling of `InvalidAccessorSparseIndicesComponentTypeException`

* 1.13.3
    * add support to `GLTF.convert_images` to convert from buffers to image files. NOTE: Does not update buffer yet.
    * add support to `GLTF.convert_images` to convert from buffers to data uris. NOTE: Does not update buffer yet.
    * add accessor sparse indices bufferview check to validator 
    * fix test suite to write tmp files to tmp directory instead of install directory
    * remove support for old versions of dataclasses-json
    * renamed CUBICSPLINE, LINEAR and STEP to ANIM_LINEAR, ANIM_STEP, ANIM_CUBICSPLINE to fix clash with webGL constants 
    * change bin file conversion to only lose binary blob on successful file export

* 1.13.2 
    * add deprecated to pip setup

* 1.13.1
    * add `GLTF.convert_images` method for converting texture image data from data uris to files and vice versa
    * add 'name' attribute to `Image`
    * add more unittests
    * change `Primitive` so that `Attributes` is created on init 

* 1.13.0
    * NOTE: There are a few small deprecations in this version to tighten up the library that will be removed in version 2.0.0
    * deprecate class `SparseAccessor` in favour of `AccessorSparseIndices` and `AccessorSparseValues` (please update your code)
    * deprecate class `MaterialTexture` in favour of `TextureInfo` to better match GLTF2 specification (please update your code)
    * deprecate `AlphaMode` ENUM in favour of constants (eg replace `AlphaMode.OPAQUE` with `OPAQUE` (please update your code)
    * fix support for `material.occlusionTextureInfo`  
    * fix support for `material.normalTextureInfo`
    * fix sampler support in `Animation` class by adding `AnimationSampler`
    * add default values for Accessor, AnimationSampler, BufferView, Material, PbrMetallicRoughness, Primitive, Sampler
    * add Optional to attributes for better type hinting
    * add initial `utils.validator` to validate GLTF2 object accessor and bufferViews
    
* 1.12.0
    * fix bug with binfile path handling
    
* 1.11.10
    * convert load methods from staticmethod to classmethods

* 1.11.9

pygltflib/utils.py
"""
pygltflib.utils.py : A collection of functions for manipulating GLTF2 objects.


Copyright (c) 2018, 2019 Luke Miller

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
"""

import base64
from struct import calcsize
import pathlib
import warnings


from . import *


# some higher level helper functions


def add_node(gltf, node):
    warnings.warn("pygltf.utils.add_node is a provisional function and may not exist in future versions.")
    if gltf.scene is not None:
        gltf.scenes[gltf.scene].nodes.append(len(gltf.nodes))
    gltf.nodes.append(node)
    return gltf


def find_node_index_by_name(gltf: GLTF2, name):
    for index, node in enumerate(gltf.nodes):
        if node.name == name:
            return index
    return -1

def add_default_camera(gltf):
    warnings.warn("pygltf.utils.add_default_camera is a provisional function and may not exist in future versions.")
    n = Node()
    n.rotation = [0.0, 0.0, 0.0, 1]
    n.translation = [-1.0, 0.0, 0.0]
    n.name = "Camera"
    n.camera = len(gltf.cameras)

    gltf.add_node(n)
    c = Camera()
    c.type = PERSPECTIVE
    c.perspective = Perspective()
    c.perspective.aspectRatio = 1.5
    c.perspective.yfov = 0.6
    c.perspective.zfar = 1000
    c.perspective.znear = 0.001
    gltf.cameras.append(c)
    return gltf


def add_default_scene(gltf):
    warnings.warn("pygltf.utils.add_default_scene is a provisional function and may not exist in future versions.")
    s = Scene()
    s.name = "Scene"
    gltf.scene = 0
    gltf.scenes.append(s)
    return gltf


def add_camera(gltf, rotation, translation, scale):
    warnings.warn("pygltf.utils.add_camera is a provisional function and may not exist in future versions.")
    n = Node()
    n.rotation = rotation
    n.translation = translation
    n.scale = scale
    n.name = "Camera"
    n.camera = len(gltf.cameras)

    gltf.add_node(n)
    c = Camera()
    c.type = PERSPECTIVE
    c.perspective = Perspective()
    c.perspective.aspectRatio = 1.5
    c.perspective.yfov = 0.6
    c.perspective.zfar = 1000
    c.perspective.znear = 0.001
    gltf.cameras.append(c)
    return gltf


def uri2vectors(uri):
    base64.b64decode(uri[len('data:application/octet-stream;base64,'):])


def indices_and_vertices_to_gltf(gltf, indices, vertices):
    pass


def get_accessor_for_bufferview(gltf, bufferview=0):
    warnings.warn("pygltf.utils.get_accessor_for_bufferview is a provisional function and may not exist in future versions.")
    for accessor in gltf.accessors:
        if accessor.bufferView == bufferview:
            return accessor
    return None

def get_bufferview_for_accessor(gltf, accessor):
    warnings.warn("pygltf.utils.get_accessor_for_bufferview is a provisional function and may not exist in future versions.")
    #bufferview =
    for bufferview in gltf.accessors:
        if accessor.bufferView == bufferview:
            return accessor
    return None


def unpackURI(gltf, buffer_index=0):
    """ Unpack a data uri of a primitive with indexed geometry

    Args:
        gltf (GLTF2): a gltf object containing indexed geometry
        buffer_index: the index pointing to the buffer to unpack in gltf.buffers

    Returns:
        indices, vertices (List(Any)): List of indices that point to elements in the list of vertices (also returned)
        """

    warnings.warn("pygltf.utils.unpackURI is a provisional function and may not exist in future versions.")

    start = 'data:application/octet-stream;base64,'
    buffer = gltf.buffers[buffer_index]
    if not buffer.uri.startswith(start):
        warnings.warn(f"buffer {buffer_index} does not appear to be a data uri.")
        return {}
    data = base64.b64decode(buffer.uri[len(start):])
    if len(data) != buffer.byteLength:
        warnings.warn(f"buffer {buffer_index} is not the expected length.")
    indices = []
    vertices = []
    for i, accessor in enumerate(gltf.accessors):
        buffer_view = gltf.bufferViews[accessor.bufferView]
        chunk = data[buffer_view.byteOffset:buffer_view.byteOffset + buffer_view.byteLength]
        if buffer_view.buffer != buffer_index:
            continue
        supported = {
            UNSIGNED_SHORT: "H",
            FLOAT: "f",

        }
        unpack = supported[accessor.componentType] * accessor.count  # one record in this data set
        size = calcsize(unpack)  # size of one record
        num_of_vals = buffer_view.byteLength//size  # num of records in this chunk
        for j in range(0, num_of_vals):
            v = struct.unpack(unpack, chunk[j*size:(j*size)+size])
            if buffer_view.target == ELEMENT_ARRAY_BUFFER:  # index data (unsigned shorts)
                indices.append(v)
            elif buffer_view.target == ARRAY_BUFFER:  # vertex data (floats)
                vertices.append(v)
            else:
                warnings.warn(f"bufferview {i} doesn't seem relevant to indexed vertices")
    return indices, vertices


def add_indexed_geometry(gltf, indices, vertices):
    """
    Add a primitive object to the GLTF that is a list of indices and vertices.
    eg a triangle with indices [(0, 1, 2)] and vertices [(0.0, 0.0, 0.0), (1.0, 0.0, 0.0), (0.0, 1.0, 0.0)]
    """
    buffer = Buffer()
    bufferView1 = BufferView()  # indices buffer view
    bufferView2 = BufferView()  # vertices buffer view
    accessor1 = Accessor()
    accessor2 = Accessor()

    mesh = Mesh()
    primitive = Primitive()
    node = Node()

    # add to gltf
    gltf.meshes.append(mesh)
    gltf.meshes[-1].primitives.append(primitive)
    gltf.nodes.append(node)
    gltf.buffers.append(buffer)
    gltf.bufferViews.append(bufferView1)
    gltf.bufferViews.append(bufferView2)
    gltf.accessors.append(accessor1)
    gltf.accessors.append(accessor2)

    bufferview1_index = len(gltf.bufferViews)-2
    bufferview2_index = len(gltf.bufferViews)-1
    buffer_index = len(gltf.buffers) - 1
    node_index = len(gltf.nodes)-1

    # accessor for indices
    accessor1.bufferView = bufferview1_index
    accessor1.byteOffset = 0
    accessor1.componentType = UNSIGNED_SHORT
    accessor1.count = 3
    accessor1.type = SCALAR
    accessor1.max = [2]
    accessor1.min = [0]

    # accessor for vertices
    accessor2.bufferView = bufferview2_index
    accessor2.byteOffset = 0
    accessor2.componentType = FLOAT
    accessor2.count = 3
    accessor2.type = VEC3
    accessor2.max = [1.0, 1.0, 0.0]
    accessor2.min = [0.0, 0.0, 0.0]

    primitive.attributes = Attributes()
    primitive.attributes.POSITION = 1
    node.mesh = 0
    scene = None
    if not gltf.scenes:
        warnings.warn("Adding primitive to GLTF but there is no scene. You may want to add one.")
    if len(gltf.scenes)>1:
        warnings.warn("Multiple scenes found, adding to most recent one.")
        scene = gltf.scenes[-1]

    if scene:
        if not scene.nodes:
            scene.nodes = [node_index]
        else:
            scene.nodes.append(node_index)

    # add the data
    stream = "data:application/octet-stream;base64,"
    buffer.uri = stream  # first part of the datastream is set up

    chunk = b""
    pack = "<HHH"
    for v in indices:
        chunk += struct.pack(pack, *v)

    bufferView1.buffer = buffer_index
    bufferView1.byteOffset = 0
    byte_length = bufferView1.byteLength = len(chunk)
    bufferView1.target = ELEMENT_ARRAY_BUFFER
    buffer.uri += base64.b64encode(chunk).decode("utf-8")  # add to data stream

    # DH: we do not need this line.
    # byte_length += 4 - byte_length % 4  # pad to next chunk

    chunk = b""
    pack = "<fff"
    for v in vertices:
        chunk += struct.pack(pack, *v)

    # record_size = byte_length * num_of_fields
    bufferView2.buffer = buffer_index
    bufferView2.byteOffset = byte_length
    bufferView2.byteLength = len(chunk)
    bufferView2.target = ARRAY_BUFFER
    buffer.uri += base64.b64encode(chunk).decode("utf-8")  # add vertices to data stream

    buffer.byteLength = bufferView2.byteOffset + bufferView2.byteLength
    return True


def add_primitive(_gltf):
    warnings.warn("pygltf.utils.add_primitive is a provisional function and may not exist in future versions.")

    # create gltf objects for a scene with a primitive triangle with indexed geometery
    gltf = GLTF2()
    scene = Scene()
    mesh = Mesh()
    primitive = Primitive()
    node = Node()
    buffer = Buffer()
    bufferView1 = BufferView()
    bufferView2 = BufferView()
    accessor1 = Accessor()
    accessor2 = Accessor()

    # add data
    primitive.attributes = Attributes()
    primitive.attributes.POSITION = 1

    buffer.uri = "data:application/octet-stream;base64,AAABAAIAAAAAAAAAAAAAAAAAAAAAAIA/AAAAAAAAAAAAAAAAAACAPwAAAAA="
    buffer.byteLength = 44

    bufferView1.buffer = 0
    bufferView1.byteOffset = 0
    bufferView1.byteLength = 6
    bufferView1.target = ELEMENT_ARRAY_BUFFER

    bufferView2.buffer = 0
    bufferView2.byteOffset = 8
    bufferView2.byteLength = 36
    bufferView2.target = ARRAY_BUFFER

    accessor1.bufferView = 0
    accessor1.byteOffset = 0
    accessor1.componentType = UNSIGNED_SHORT
    accessor1.count = 3
    accessor1.type = SCALAR
    accessor1.max = [2]
    accessor1.min = [0]

    accessor2.bufferView = 1
    accessor2.byteOffset = 0
    accessor2.componentType = FLOAT
    accessor2.count = 3
    accessor2.type = VEC3
    accessor2.max = [1.0, 1.0, 0.0]
    accessor2.min = [0.0, 0.0, 0.0]

    node.mesh = 0
    scene.nodes = [0]

    # assemble into a gltf structure
    gltf.scenes.append(scene)
    gltf.meshes.append(mesh)
    gltf.meshes[0].primitives.append(primitive)
    gltf.nodes.append(node)
    gltf.buffers.append(buffer)
    gltf.bufferViews.append(bufferView1)
    gltf.bufferViews.append(bufferView2)
    gltf.accessors.append(accessor1)
    gltf.accessors.append(accessor2)
    # save to file
    # gltf.save("primitive.glb")
    unpackURI(gltf, 0)
    return gltf


def gltf2glb(source, destination=None, override=False):
    """
    Save a .gltf file as its .glb equivalent.

    Args:
        source (str): Path to existing .gltf file.
        destination (Optional(str)): Filename to write to (default is to use existing filename as base)
        override (bool): Override existing file.

    """
    path = Path(source)
    if not destination:
        destination = path.with_suffix(".glb")
    else:
        destination = Path(destination)
    if destination.is_file() and override is False:
        raise FileExistsError
    else:
        GLTF2().load(str(path)).save_binary(str(destination))
    return True


def glb2gltf(source, destination=None, override=False):
    """
    Save a .glb file as its .gltf equivalent.

    Args:
        source (str): Path to existing .glb file.
        destination (Optional(str)): Filename to write to (default is to use existing filename as base)
        override (bool): Override existing file.

    """
    path = Path(source)
    if not destination:
        destination = path.with_suffix(".gltf")
    else:
        destination = Path(destination)
    if destination.is_file() and override is False:
        raise FileExistsError
    else:
        GLTF2().load(str(path)).save_json(str(destination))
    return True

